File: packages/core/src/protocol/timestamps.ts
import type { Timestamps } from "./types.js"

/**
 * Returns true if the value is a "leaf" in our data model
 * (anything that is not a plain object — primitives, arrays, null).
 */
function isLeaf(v: unknown): boolean {
  if (v === null || v === undefined) return true
  if (Array.isArray(v)) return true
  return typeof v !== "object"
}

/**
 * Compute new timestamps by diffing old and new data trees.
 * - Unchanged leaf values keep their old timestamp
 * - Changed or new values get `now`
 * - Removed keys are omitted from the result
 */
export function computeTimestamps(
  oldData: Record<string, unknown> | null,
  newData: Record<string, unknown>,
  oldTimestamps: Timestamps | null,
  now: number
): Timestamps {
  const result: Timestamps = {}

  for (const key of Object.keys(newData)) {
    const newVal = newData[key]
    const oldVal = oldData?.[key]
    const oldTs = oldTimestamps?.[key]

    if (isLeaf(newVal)) {
      // Leaf: compare with old value
      if (oldData && key in oldData && isLeaf(oldVal) && stableEqual(oldVal, newVal) && typeof oldTs === "number") {
        result[key] = oldTs
      } else {
        result[key] = now
      }
    } else {
      // Object: recurse
      const newObj = newVal as Record<string, unknown>
      const oldObj = (!isLeaf(oldVal) && oldVal !== null && oldVal !== undefined)
        ? oldVal as Record<string, unknown>
        : null
      const oldTsObj = (oldTs !== null && oldTs !== undefined && typeof oldTs === "object")
        ? oldTs as Timestamps
        : null
      result[key] = computeTimestamps(oldObj, newObj, oldTsObj, now)
    }
  }

  return result
}

/**
 * Filter data to only include paths where the timestamp > checkpoint.
 * Maintains nesting structure.
 */
export function filterByCheckpoint(
  data: Record<string, unknown>,
  timestamps: Timestamps,
  checkpoint: number
): Record<string, unknown> {
  const result: Record<string, unknown> = {}

  for (const key of Object.keys(data)) {
    const val = data[key]
    const ts = timestamps[key]

    if (ts === undefined) continue

    if (typeof ts === "number") {
      // Leaf timestamp
      if (ts > checkpoint) {
        result[key] = val
      }
    } else {
      // Nested object
      if (isLeaf(val)) {
        // Mismatch: timestamps say object but data is leaf — include if any sub-ts > checkpoint
        result[key] = val
      } else {
        const filtered = filterByCheckpoint(
          val as Record<string, unknown>,
          ts,
          checkpoint
        )
        if (Object.keys(filtered).length > 0) {
          result[key] = filtered
        }
      }
    }
  }

  return result
}

/**
 * Deep equality for leaf values (primitives, arrays, null).
 */
function stableEqual(a: unknown, b: unknown): boolean {
  if (a === b) return true
  if (a === null || b === null) return a === b
  if (typeof a !== typeof b) return false
  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) return false
    for (let i = 0; i < a.length; i++) {
      if (!stableEqual(a[i], b[i])) return false
    }
    return true
  }
  return false
}

File: servers/python/satellite_server/protocol/timestamps.py
"""Timestamp computation and checkpoint filtering."""

from __future__ import annotations

import json
from typing import Any

from satellite_server.protocol.types import Timestamps


def _is_leaf(v: Any) -> bool:
    """Return True if the value is a leaf (not a plain dict)."""
    if v is None:
        return True
    if isinstance(v, list):
        return True
    return not isinstance(v, dict)


def _stable_equal(a: Any, b: Any) -> bool:
    """Deep equality for leaf values (primitives, arrays, None)."""
    if a is b:
        return True
    if a is None or b is None:
        return a is b
    if type(a) is not type(b):
        return False
    if isinstance(a, list) and isinstance(b, list):
        if len(a) != len(b):
            return False
        return all(_stable_equal(x, y) for x, y in zip(a, b))
    return a == b


def compute_timestamps(
    old_data: dict[str, Any] | None,
    new_data: dict[str, Any],
    old_timestamps: Timestamps | None,
    now: int,
) -> Timestamps:
    """Compute new timestamps by diffing old and new data trees.

    - Unchanged leaf values keep their old timestamp
    - Changed or new values get ``now``
    - Removed keys are omitted from the result
    """
    result: Timestamps = {}

    for key in new_data:
        new_val = new_data[key]
        old_val = old_data.get(key) if old_data else None
        old_ts = old_timestamps.get(key) if old_timestamps else None

        if _is_leaf(new_val):
            if (
                old_data is not None
                and key in old_data
                and _is_leaf(old_val)
                and _stable_equal(old_val, new_val)
                and isinstance(old_ts, int)
            ):
                result[key] = old_ts
            else:
                result[key] = now
        else:
            # Object: recurse
            new_obj = new_val
            old_obj = old_val if (not _is_leaf(old_val) and old_val is not None) else None
            old_ts_obj = old_ts if isinstance(old_ts, dict) else None
            result[key] = compute_timestamps(old_obj, new_obj, old_ts_obj, now)

    return result


def filter_by_checkpoint(
    data: dict[str, Any],
    timestamps: Timestamps,
    checkpoint: int,
) -> dict[str, Any]:
    """Filter data to only include paths where the timestamp > checkpoint."""
    result: dict[str, Any] = {}

    for key in data:
        val = data[key]
        ts = timestamps.get(key)

        if ts is None:
            continue

        if isinstance(ts, int):
            if ts > checkpoint:
                result[key] = val
        else:
            # Nested object timestamps
            if _is_leaf(val):
                # Mismatch: timestamps say object but data is leaf
                result[key] = val
            else:
                filtered = filter_by_checkpoint(val, ts, checkpoint)
                if filtered:
                    result[key] = filtered

    return result
